### PR TITLE
Condense difficulty icons when needed in the listing page

### PR DESCRIPTION
**Pull request by flesnuk**
If there are more than 10 diffs in the beatmap listing, condense them showing for each mode the icon of the hardest diff and the number of diffs of that mode.

### CODE REVIEW

File: resources/assets/less/bem/beatmapset-panel.less [hunk 1]
```diff
@@ -57,6 +57,12 @@
     margin-right: 5px;
   }
 
+  &__difficulty-count {
+    margin-right: 8px;
```
Comment (nanaya) [inline resources/assets/less/bem/beatmapset-panel.less]: need a bit more margin (3px left, 10px right)

File: resources/assets/coffee/react/_components/beatmapset-panel.coffee [hunk 2]
```diff
@@ -50,16 +50,27 @@ class @BeatmapsetPanel extends React.PureComponent
...
-            key: b.id
-            el BeatmapIcon, beatmap: b
-
-    if beatmapset.beatmaps.length > maxDisplayedDifficulty
-      difficulties.push span key: 'over', "+#{(beatmapset.beatmaps.length - maxDisplayedDifficulty)}"
+      for own _mode, beatmaps of BeatmapHelper.group beatmapset.beatmaps
+        if condenseDifficulties
+          [..., hardestDiff] = beatmaps
+          [ 
+            beatmapIcon hardestDiff
```
Comment (nanaya) [inline resources/assets/coffee/react/_components/beatmapset-panel.coffee]: should not show title tooltip (`showTitle: false`).

Comment (nanaya) [inline resources/assets/coffee/react/_components/beatmapset-panel.coffee]: skipping `beatmapIcon` and directly calling `el BeatmapIcon` should work.

File: resources/assets/coffee/react/_components/beatmapset-panel.coffee [hunk 3]
```diff
@@ -50,16 +50,27 @@ class @BeatmapsetPanel extends React.PureComponent
...
-            className: 'beatmapset-panel__difficulty-icon'
-            key: b.id
-            el BeatmapIcon, beatmap: b
-
-    if beatmapset.beatmaps.length > maxDisplayedDifficulty
-      difficulties.push span key: 'over', "+#{(beatmapset.beatmaps.length - maxDisplayedDifficulty)}"
+      for own _mode, beatmaps of BeatmapHelper.group beatmapset.beatmaps
+        if condenseDifficulties
+          [..., hardestDiff] = beatmaps
+          [
```
Comment (nanaya) [inline resources/assets/coffee/react/_components/beatmapset-panel.coffee]: can also use `span` here and use mode as `key` so the containing icon and count don't need key.

**Comment by flesnuk** [comment]
done

File: resources/assets/coffee/react/_components/beatmapset-panel.coffee [hunk 4]
```diff
@@ -50,16 +50,25 @@ class @BeatmapsetPanel extends React.PureComponent
...
-            el BeatmapIcon, beatmap: b
-
-    if beatmapset.beatmaps.length > maxDisplayedDifficulty
-      difficulties.push span key: 'over', "+#{(beatmapset.beatmaps.length - maxDisplayedDifficulty)}"
+      for own mode, beatmaps of BeatmapHelper.group beatmapset.beatmaps
+        if condenseDifficulties
+          [..., hardestDiff] = beatmaps
+          span
+            key: mode
+            style:
```
Comment (nanaya) [inline resources/assets/coffee/react/_components/beatmapset-panel.coffee]: hahaha this looks terrible. Maybe keep the array after all.

Comment (flesnuk) [inline resources/assets/coffee/react/_components/beatmapset-panel.coffee]: Lol, I don't know which is worse, but yeah, with the array looked cleaner :neutral_face:

File: resources/assets/coffee/react/_components/beatmapset-panel.coffee [hunk 5]
```diff
@@ -50,16 +50,27 @@ class @BeatmapsetPanel extends React.PureComponent
...
-      for own _mode, beatmaps of BeatmapHelper.group beatmapset.beatmaps[..maxDisplayedDifficulty - 1]
-        for b in beatmaps
-          div
-            className: 'beatmapset-panel__difficulty-icon'
-            key: b.id
-            el BeatmapIcon, beatmap: b
-
-    if beatmapset.beatmaps.length > maxDisplayedDifficulty
-      difficulties.push span key: 'over', "+#{(beatmapset.beatmaps.length - maxDisplayedDifficulty)}"
+      for own mode, beatmaps of BeatmapHelper.group beatmapset.beatmaps
```
Comment (nanaya) [inline resources/assets/coffee/react/_components/beatmapset-panel.coffee]: a combination of previous span and this is slightly better, I think. Also, `hardestDiff` can be inlined with `_.last(beatmaps)`.

```coffeescript
if condenseDifficulties
  [
    el BeatmapIcon, key: "#{mode}-icon", beatmap: _.last(beatmaps), showTitle: false
    span
      className: 'beatmapset-panel__difficulty-count'
      key: "#{(mode)}-count", beatmaps.length
  ]
else
  for b in beatmaps
    div
      className: 'beatmapset-panel__difficulty-icon'
      key: b.id
      el BeatmapIcon, beatmap: b
```

Comment (flesnuk) [inline resources/assets/coffee/react/_components/beatmapset-panel.coffee]: Didn't even think about lodash, thanks. I still have to learn a lot.

**Pull request merged**